source/MapPanel.cpp

```diff
@@ -1378,30 +1378,32 @@ void MapPanel::DrawSystems()
 	if(commodity != cachedCommodity)
 		UpdateCache();
 
-	// If coloring by government, we need to keep track of which ones are the
-	// closest to the center of the window because those will be the ones that
-	// are shown in the map key.
+	// If coloring by government, we need to keep track of which ones are
+	// most prevalent.
 	if(commodity == SHOW_GOVERNMENT)
-		closeGovernments.clear();
+		governmentCounts.clear();
 
 	// Draw the circles for the systems.
 	double zoom = Zoom();
+	hasUnexplored = false;
+	hasUninhabited = false;
 	for(const Node &node : nodes)
 	{
 		Point pos = zoom * (node.position + center);
 		RingShader::Draw(pos, OUTER, INNER, node.color);
 
-		if(commodity == SHOW_GOVERNMENT && node.government && node.government->GetName() != "Uninhabited")
+		// Count the system for the government (to choose what government labels to display).
+		// Make sure that it is in an oval that is stretched to fit in the centers of all the edges
+		// before counting it.
+		if(commodity == SHOW_GOVERNMENT &&
+			(pos / Screen::BottomRight()).LengthSquared() < 1)
```

Comment:
It might be easier to construct a Rectangle describing the screen before the loop, and check if that contains the system. It would be more accurate, too. (And you wouldn't need the Point changes.)